builder/openstack/step_discover_network.go

```diff
@@ -0,0 +1,56 @@
+package openstack
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/hashicorp/packer/helper/multistep"
+	"github.com/hashicorp/packer/packer"
+)
+
+type StepDiscoverNetwork struct {
+	Networks              []string
+	NetworkDiscoveryCIDRs []string
+	Ports                 []string
+}
+
+func (s *StepDiscoverNetwork) Run(ctx context.Context, state multistep.StateBag) multistep.StepAction {
+	config := state.Get("config").(*Config)
+	ui := state.Get("ui").(packer.Ui)
+
+	networkClient, err := config.networkV2Client()
+	if err != nil {
+		err = fmt.Errorf("Error initializing network client: %s", err)
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+
+	networks := make([]servers.Network, len(s.Networks)+len(s.Ports))
+	i := 0
+	for ; i < len(s.Ports); i++ {
+		networks[i].Port = s.Ports[i]
+	}
+	for ; i < len(networks); i++ {
+		networks[i].UUID = s.Networks[i-len(s.Ports)]
+	}
```

Comment:
Nit: this part looks complicated and could be simpler:

Suggestion:
```go
	networks := []servers.Network{} // or make([]servers.Network, 0, len(s.Networks)+len(s.Ports)) if we really wanted to optimize memory here, but I don't think this is necessary given what's after line 38
	for _, port := range s.Ports {
		networks = append(networks, servers.Network{Port : port})
	}
	for _, uuid := range s.Networks {
		networks = append(networks, servers.Network{UUID : uuid})
	}
```

Reply:
Thank you. Updated it :-)